src/main/cpp/src/ftos_converter.cu

```diff
@@ -0,0 +1,1569 @@
+            sep_cnt = 0;
+        }
+      }
+      for (int i = 0; i < olength; i++) {
+        if (sep_cnt == 3) {
+          result[integer_len - (rev_index++) - 1] = ',';
+          sep_cnt = 0;
+        }
+        result[integer_len - (rev_index++) - 1] = (char) ('0' + output % 10);
+        sep_cnt++;
+        output /= 10;
+      }
+      index = integer_len;
+      if (d == 0) {
+        return index;
+      }
+      result[index++] = '.';
+      for (int i = 0; i < d; i++) {
+        result[index++] = '0';
+      }
+    } else {
+      uint32_t temp_d = d, tailing_zero = 0;
+      if (exp + d > olength) {
+        temp_d = olength - exp;
+        tailing_zero = d - temp_d;
+      }
+      uint32_t rounded_output = round_half_even(output, olength, exp+temp_d+1);
+      uint32_t pow10 = POW10_TABLE[temp_d];
+      uint32_t integer = rounded_output / pow10;
+      uint32_t decimal = rounded_output % pow10;
+      // calculate integer length after format to cover carry case
+      uint32_t integer_len = decimalLength9(integer);
+      uint32_t formated_integer_len = index + integer_len + (integer_len - 1) / 3;
+      uint32_t sep_cnt = 0;
+      int rev_index = 0;
+      for (int i = 0; i < integer_len; i++) {
+        if (sep_cnt == 3) {
+          result[formated_integer_len - (rev_index++) - 1] = ',';
+          sep_cnt = 0;
+        }
+        result[formated_integer_len - (rev_index++) - 1] = (char) ('0' + integer % 10);
+        sep_cnt++;
+        integer /= 10;
+      }
+      index = formated_integer_len;
+      if (d == 0) {
+        return index;
+      }
+      result[index++] = '.';
+      int current = index;
+      for (int i = 0; i < tailing_zero; i++) {
+        result[current + d - i - 1] = '0';
+        index++;
+      }
+      for (int i = tailing_zero; i < d; i++) {
+        result[current + d - i - 1] = (char) ('0' + decimal % 10);
+        decimal /= 10;
+        index++;
+      }
+    }
+    return index;
+  }
+
+  __device__ inline int format_float_size(const floating_decimal_32 v, const bool sign, int d) {
+    int index = 0;
+    if (sign) {
+      index++;
+    }
+    uint64_t output = v.mantissa;
+    const uint32_t olength = decimalLength9(output);
+    int32_t exp = v.exponent + (int32_t) olength - 1;
+    if (exp < 0) {
+      index += 2 + d;
+    } else if (exp + 1 >= olength) {
+      index += exp + 1 + exp / 3 + 1 + d;
+    } else {
+      uint32_t temp_d = d;
+      if (exp + d > olength) {
+        temp_d = olength - exp;
+      }
+      uint64_t rounded_output = round_half_even(output, olength, exp+temp_d+1);
+      uint64_t pow10 = POW10_TABLE[temp_d];
+      uint64_t integer = rounded_output / pow10;
+      uint32_t integer_len = decimalLength9(integer);
+      index += integer_len + (integer_len - 1) / 3 + 1 + d;
+    }
+    if (d == 0) {
+      index--;
+    }
+    return index;
+  }  
+
+  __device__ int compute_format_float_size(double value, int d, bool is_float) {
```

Comment:
So this and `format_float` are the only functions of interest. So just add declaration for them in a new header (like `ftos_converter.cuh`).
```
__device__ int compute_format_float_size(double value, int d, bool is_float);
__device__ int format_float(double value, int d, char* output, bool is_float);

```

Reply:
And since all other functions don't need to be exposed to other files, just wrap them in the anonymous namespace. Only two functions above are in the public namespace:
```
namespace spark_rapids_jni::ftos_converter {
namespace {
....
} // namespace

__device__ int compute_format_float_size(double value, int d, bool is_float) {...}
__device__ int format_float(double value, int d, char* output, bool is_float) {...}

} // namespace spark_rapids_jni::ftos_converter
```

Please also add comment like `// namespace` or `// namespace spark_rapids_jni::ftos_converter` so we can know where a namespace end.

Reply:
I put the whole thing in a .cuh file, not sure if it's good.

Reply:
It is better to split into `.cuh` and `.cu` files so we can have `ftos_converter.cu` and `format_float.cu` compiled in parallel.